# 1 Perform a set of exercises. All tasks should be performed in two ways:

## (A)using universal functions of the NumPy library

In [7]:
#task1
import numpy as np 

a = np.zeros((5, 5), int);
np.fill_diagonal((a), [1,2,3,4,5])
a

array([[1, 0, 0, 0, 0],
       [0, 2, 0, 0, 0],
       [0, 0, 3, 0, 0],
       [0, 0, 0, 4, 0],
       [0, 0, 0, 0, 5]])

In [10]:
#task3
import numpy as np 

a = np.zeros((6,7), int);
b = 1 #easy access to change value 
a[1,1] = b
a[1,2] = b
a[1,4] = b
a[1,5] = b
a[3,1] = b
a[3,2] = b
a[3,4] = b
a[3,5] = b
a[5,1] = b
a[5,2] = b
a[5,4] = b
a[5,5] = b
a

array([[0, 0, 0, 0, 0, 0, 0],
       [0, 1, 1, 0, 1, 1, 0],
       [0, 0, 0, 0, 0, 0, 0],
       [0, 1, 1, 0, 1, 1, 0],
       [0, 0, 0, 0, 0, 0, 0],
       [0, 1, 1, 0, 1, 1, 0]])

In [22]:
#task4
import numpy as np

a = np.zeros((4,3), int);
a[0,1] = 1
a[0,2] = 2
a

array([[0, 1, 2],
       [0, 0, 0],
       [0, 0, 0],
       [0, 0, 0]])

In [27]:
#task6
import numpy as np

ini_array1 = np.array([3, 4, 0, 9, 7, 0, 6, 0, 4, 0, 3])

print("initial array", ini_array1)

ini_array1[ini_array1>0] = -1

print("New resulting array: ", ini_array1)

initial array [3 4 0 9 7 0 6 0 4 0 3]
New resulting array:  [-1 -1  0 -1 -1  0 -1  0 -1  0 -1]


In [48]:
#task10
import numpy as np

a = np.random.rand(3,3)

min = np.min(a)
max = np.max(a)
mean = np.mean(a)
std = np.std(a)

mmin = np.round(min, 3)
mmax = np.round(max,3)
mmean = np.round(mean,3)
sstd = np.round(std,3)

print("Array =", a)
print("Mini =", mmin)
print("Max =", mmax)
print("Mean =", mmean)
print("Standard Deviation =", sstd)

Array = [[0.91851217 0.86305177 0.12825092]
 [0.37266766 0.73863274 0.20914769]
 [0.27414737 0.9932725  0.26994428]]
Mini = 0.128
Max = 0.993
Mean = 0.53
Standard Deviation = 0.324


In [66]:
#task12
import numpy as np
x = np.ones((2,2))
print("Original array:")
print(x)
print("0 on the border and 1 inside in the array")
x = np.pad(x, pad_width=1, mode='constant', constant_values=0)
print(x)

Original array:
[[1. 1.]
 [1. 1.]]
0 on the border and 1 inside in the array
[[0. 0. 0. 0.]
 [0. 1. 1. 0.]
 [0. 1. 1. 0.]
 [0. 0. 0. 0.]]


In [129]:
#task15
import numpy as np

a = np.zeros((5,5))
o = a
o = a[:,0::2]
e = a
e = a[:,1::2]
e.fill(1)
print(a)

[[0. 1. 0. 1. 0.]
 [0. 1. 0. 1. 0.]
 [0. 1. 0. 1. 0.]
 [0. 1. 0. 1. 0.]
 [0. 1. 0. 1. 0.]]


In [34]:
#task17
import numpy as np 

a = np.arange(11)
G = np.where(a < (11/2), a, -a)
L = np.where(a > ((3 * 11)/4), a, -a)
c = np.stack((G,L),axis=1)

print("less than n/2 =", L)
print("greater than 3n/4", G)
print(c)

less than n/2 = [ 0 -1 -2 -3 -4 -5 -6 -7 -8  9 10]
greater than 3n/4 [  0   1   2   3   4   5  -6  -7  -8  -9 -10]
[[  0   0]
 [  1  -1]
 [  2  -2]
 [  3  -3]
 [  4  -4]
 [  5  -5]
 [ -6  -6]
 [ -7  -7]
 [ -8  -8]
 [ -9   9]
 [-10  10]]


In [9]:
#task18
import numpy as np

a = np.random.randint(99, size=(30))
(unique, counts) = np.unique(a,return_counts=True)

frequencies = np.asarray((unique, counts)).T

print("Unique Values and their Frequency list",frequencies)

Unique Values and their Frequency list [[ 0  1]
 [ 4  1]
 [14  1]
 [16  1]
 [24  1]
 [26  1]
 [37  1]
 [45  1]
 [46  1]
 [48  2]
 [50  2]
 [51  2]
 [66  1]
 [70  1]
 [71  1]
 [75  1]
 [76  1]
 [77  1]
 [80  1]
 [81  1]
 [88  1]
 [92  1]
 [93  2]
 [96  3]]


In [35]:
#(A)calculate the execution time of the script.

%timeit sum(range(100))

1e+03 ns ± 17.5 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


## (B)using iterative constructions (using loops, list comprehensions, etc.)

In [61]:
#task1

import numpy as np 

a = [[1,0,0,0,0],[0,2,0,0,0],[0,0,3,0,0],[0,0,0,4,0],[0,0,0,0,5]]
a

[[1, 0, 0, 0, 0],
 [0, 2, 0, 0, 0],
 [0, 0, 3, 0, 0],
 [0, 0, 0, 4, 0],
 [0, 0, 0, 0, 5]]

In [64]:
#task3

import numpy as np 

a = np.zeros((6,7), int);
b = 1 #easy access to change value 
a[1,1] = b
a[1,2] = b
a[1,4] = b
a[1,5] = b
a[3,1] = b
a[3,2] = b
a[3,4] = b
a[3,5] = b
a[5,1] = b
a[5,2] = b
a[5,4] = b
a[5,5] = b
a

array([[0, 0, 0, 0, 0, 0, 0],
       [0, 1, 1, 0, 1, 1, 0],
       [0, 0, 0, 0, 0, 0, 0],
       [0, 1, 1, 0, 1, 1, 0],
       [0, 0, 0, 0, 0, 0, 0],
       [0, 1, 1, 0, 1, 1, 0]])

In [62]:
#task4

a = [[0, 1, 2],
       [0, 0, 0],
       [0, 0, 0],
       [0, 0, 0]]
a

[[0, 1, 2], [0, 0, 0], [0, 0, 0], [0, 0, 0]]

In [49]:
#task6

a = [3, 4, 0, 9, 7, 0, 6, 0, 4, 0, 3]
[-1 if i > 0 else i for i in a]

[-1, -1, 0, -1, -1, 0, -1, 0, -1, 0, -1]

In [72]:
#task10

import numpy as np
import statistics

randnums= np.random.randint(1,26,10)

mean = sum(randnums) / len(randnums)

std = statistics.stdev(randnums)

print("Array", randnums)
print("Max value element : ", max(randnums))
print("Min value element : ", min(randnums))
print("Mean value element:", mean)
print("Standard Deviation value:", std)

Array [ 4 12 25 15 21 25 20 11 22 17]
Max value element :  25
Min value element :  4
Mean value element: 17.2
Standard Deviation value: 6.708203932499369


In [ ]:
#task12

import numpy as np
x = np.ones((2,2))
print("Original array:")
print(x)
print("0 on the border and 1 inside in the array")
x = np.pad(x, pad_width=1, mode='constant', constant_values=0)
print(x)

In [ ]:
#task15

import numpy as np

a = np.zeros((5,5))
o = a
o = a[:,0::2]
e = a
e = a[:,1::2]
e.fill(1)
print(a)

In [ ]:
#task17

#i dont know how to do task 17 at all, maybe im just missing something 


In [12]:
#task18
import random

a_list = [random.randrange(1, 100, 1) for i in range(30)]
a_set = set(a_list)
number_of_unique_values = len(a_set)
print(number_of_unique_values)

26


In [63]:
#(B)calculate the execution time of the script
%timeit sum(range(100))

1.01 µs ± 23.1 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


## Time Analysis
Time in Task 1(A) 1e+03 ns ± 17.5 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each),
Time in Task 1(B) 1.01 µs ± 23.1 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

# 2.Solve the system of linear equations using Cramer's formulas and check the result using:

## (A) matrix multiplication;

In [28]:
import numpy as np

a = np.array([[1, 5, 3, -4], [3, 1, -2, 0], [5, -7, 0, 10], [0, 3, -5, 0]])
b= np.array([20, 9, -9, 1])

print("A =", a)

print("b =", b)

print("Ab =",np.matmul(a,b))


A = [[ 1  5  3 -4]
 [ 3  1 -2  0]
 [ 5 -7  0 10]
 [ 0  3 -5  0]]
b = [20  9 -9  1]
Ab = [34 87 47 72]


## (B) inverse matrix;

In [29]:
import numpy as np

m_list = [[1, 5, 3, -4], [3, 1, -2, 0], [5, -7, 0, 10], [0, 3, -5, 0]]
A = np.array(m_list)

inv_A = np.linalg.inv(A)

print(inv_A)

B = np.array([20, 9, -9, 1])
X = np.linalg.inv(A).dot(B)

print(X)

[[ 0.01587302  0.31746032  0.00634921 -0.11746032]
 [ 0.23809524 -0.23809524  0.0952381   0.23809524]
 [ 0.14285714 -0.14285714  0.05714286 -0.05714286]
 [ 0.15873016 -0.32539683  0.16349206  0.22539683]]
[ 3.  2.  1. -1.]


## (C) numpy.linalg.solve().

In [57]:
import numpy as np

a = np.array([[1, 5, 3, -4], [3, 1, -2, 0], [5, -7, 0, 10], [0, 3, -5, 0]])
b = np.array([20, 9, -9, 1])
x = np.linalg.solve(a, b)
x

array([ 3.,  2.,  1., -1.])

In [63]:
# 2 Compare the results using numpy.allclose().
np.allclose([34, 87, 47, 72], [3,  2,  1, -1], [3,  2,  1, -1])

False

# 3 Calculate the value of the matrix expression

## (A) using the universal functions of the NumPy library.

In [109]:
import numpy as np

a = np.array([5, 2, 0, 10, 4, 1, 7, 3, 2])
b = np.array([3, 6, -1, -1, -2, 0, 2, 1, 3])

Bb= np.array(b ** 2)
aa= np.array(a * 2)

print("Value =", (a - Bb)*(aa + b))

Value = [ -52 -340    1  171    0    2   48   14  -49]


In [100]:
#(A)calculate the execution time of the script.
%timeit (1.0 / big_array)

4.1 ms ± 215 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


## (B) using iterative constructions (using loops, list comprehensions, etc.).

In [118]:
A = [[5, 2, 0], [10, 4, 1], [7, 3, 2]]
B = [[3, 6, -1], [-1, -2, 0], [2, 1, 3]]

result = [[sum(a*b for a,b in zip(A_row,B_col)) for B_col in zip(*B)] for A_row in A]

for r in result:
    print(r)

[13, 26, -5]
[28, 53, -7]
[22, 38, -1]


In [119]:
#(B)calculate the execution time of the script.
%timeit (1.0 / big_array)

3.96 ms ± 143 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


# 3. Calculate the value of the matrix expression:
Time in Task 3(A) 4.1 ms ± 215 µs per loop (mean ± std. dev. of 7 runs, 100 loops each), 
Time in Task 3(B) 3.96 ms ± 143 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
as shown once again iterative constructions are slightly faster can ufunc in processing speed